internal/cluster/store/store.go

```diff
@@ -0,0 +1,855 @@
+	}
+	s.boltStore = boltDB
+	logStore = s.boltStore
+	stableStore = s.boltStore
+
+	// Instantiate the Raft systems.
+	ra, err := raft.NewRaft(config, NewFSM(s), logStore, stableStore, snapshots, s.raftTn)
+	if err != nil {
+		return fmt.Errorf("new raft: %s", err)
+	}
+	s.raft = ra
+	s.observerChan = make(chan raft.Observation, observerChanLen)
+	blocking := true
+	s.observer = raft.NewObserver(s.observerChan, blocking, func(o *raft.Observation) bool {
+		_, isLeaderChange := o.Data.(raft.LeaderObservation)
+		_, isFailedHeartBeat := o.Data.(raft.FailedHeartbeatObservation)
+		_, isPeerChange := o.Data.(raft.PeerObservation)
+		return isLeaderChange || isFailedHeartBeat || isPeerChange
+	})
+	s.raft.RegisterObserver(s.observer)
+
+	s.observerClose, s.observerDone = s.observe()
+	return nil
+}
+
+// Join joins a node, identified by id and located at addr, to this store.
+// The node must be ready to respond to Raft communications at that address.
+func (s *Store) Join(jr *zproto.JoinRequest) error {
+	if !s.open.Load() {
+		return ErrNotOpen
+	}
+
+	if s.raft.State() != raft.Leader {
+		return ErrNotLeader
+	}
+
+	id := jr.Id
+	addr := jr.Address
+	voter := jr.Voter
+
+	// Confirm that this node can resolve the remote address. This can happen due
+	// to incomplete DNS records across the underlying infrastructure. If it can't
+	// then don't consider this join attempt successful -- so the joining node
+	// will presumably try again.
+	if _, err := resolvableAddress(addr); err != nil {
+		return fmt.Errorf("failed to resolve %s: %w", addr, err)
+	}
+
+	configFuture := s.raft.GetConfiguration()
+	if err := configFuture.Error(); err != nil {
+		s.logger.Info(fmt.Sprintf("failed to get raft configuration: %v", err))
+		return err
+	}
+
+	for _, srv := range configFuture.Configuration().Servers {
+		// If a node already exists with either the joining node's ID or address,
+		// that node may need to be removed from the config first.
+		if srv.ID == raft.ServerID(id) || srv.Address == raft.ServerAddress(addr) {
+			// However, if *both* the ID and the address are the same, then no
+			// join is actually needed.
+			if srv.Address == raft.ServerAddress(addr) && srv.ID == raft.ServerID(id) {
+				stats.Add(numIgnoredJoins, 1)
+				s.logger.Info(fmt.Sprintf("node %s at %s already member of cluster, ignoring join request", id, addr))
+				return nil
+			}
+
+			if err := s.remove(id); err != nil {
+				s.logger.Error(fmt.Sprintf("failed to remove node %s: %v", id, err))
+				return err
+			}
+			stats.Add(numRemovedBeforeJoins, 1)
+			s.logger.Info(fmt.Sprintf("removed node %s prior to rejoin with changed ID or address", id))
+		}
+	}
+
+	var f raft.IndexFuture
+	if voter {
+		f = s.raft.AddVoter(raft.ServerID(id), raft.ServerAddress(addr), 0, 0)
+	} else {
+		f = s.raft.AddNonvoter(raft.ServerID(id), raft.ServerAddress(addr), 0, 0)
+	}
+	if e := f.(raft.Future); e.Error() != nil {
+		if e.Error() == raft.ErrNotLeader {
```

Comment:
Comparing errors by equal sign can fail in case of error wrapping. suggesting the following:
```suggestion
		if errors.Is(e.Error(), raft.ErrNotLeader) {
```